From fdc3a239deff5213de29f487a05cf4f72e437cf6 Mon Sep 17 00:00:00 2001
From: Varun Sethi <Varun.Sethi@freescale.com>
Date: Fri, 22 Jan 2016 16:03:51 +0800
Subject: [PATCH 022/124] pci/layerscape: PCIe controller stream ID
 partitioning patch

U-boot sets up the device Id to stream ID translation table. The SMMU
driver while creating the IOMMU group programs the hardware table with
these values.

Signed-off-by: Varun Sethi <Varun.Sethi@freescale.com>
Signed-off-by: Stuart Yoder <stuart.yoder@freescale.com>
Signed-off-by: Mingkai Hu <mingkai.hu@nxp.com>
---
 drivers/pci/host/pci-layerscape.c | 42 +++++++++++++++++++++++++++++++++++++++
 drivers/pci/host/pci-layerscape.h | 19 ------------------
 2 files changed, 42 insertions(+), 19 deletions(-)

diff --git a/drivers/pci/host/pci-layerscape.c b/drivers/pci/host/pci-layerscape.c
index 3923bed..122e224 100644
--- a/drivers/pci/host/pci-layerscape.c
+++ b/drivers/pci/host/pci-layerscape.c
@@ -37,6 +37,11 @@
 
 /* PEX LUT registers */
 #define PCIE_LUT_DBG		0x7FC /* PEX LUT Debug Register */
+#define PCIE_LUT_UDR(n)		(0x800 + (n) * 8)
+#define PCIE_LUT_LDR(n)		(0x804 + (n) * 8)
+#define PCIE_LUT_MASK_ALL	0xffff
+#define PCIE_LUT_DR_NUM		32
+#define PCIE_LUT_ENABLE		(1 << 31)
 
 struct ls_pcie_drvdata {
 	u32 lut_offset;
@@ -51,10 +56,30 @@ struct ls_pcie {
 	struct pcie_port pp;
 	const struct ls_pcie_drvdata *drvdata;
 	int index;
+	const u32 *avail_streamids;
+	int streamid_index;
 };
 
 #define to_ls_pcie(x)	container_of(x, struct ls_pcie, pp)
 
+u32 set_pcie_streamid_translation(struct pci_dev *pdev, u32 devid)
+{
+	u32 index, streamid;
+	struct pcie_port *pp = pdev->bus->sysdata;
+	struct ls_pcie *pcie = to_ls_pcie(pp);
+
+	if (!pcie->avail_streamids || !pcie->streamid_index)
+		return ~(u32)0;
+
+	index = --pcie->streamid_index;
+	/* mask is set as all zeroes, want to match all bits */
+	iowrite32((devid << 16), pcie->lut + PCIE_LUT_UDR(index));
+	streamid = be32_to_cpup(&pcie->avail_streamids[index]);
+	iowrite32(streamid | PCIE_LUT_ENABLE, pcie->lut + PCIE_LUT_LDR(index));
+
+	return streamid;
+}
+
 static bool ls_pcie_is_bridge(struct ls_pcie *pcie)
 {
 	u32 header_type;
@@ -250,10 +275,27 @@ static int __init ls_pcie_probe(struct platform_device *pdev)
 
 	pcie->drvdata = match->data;
 	pcie->lut = pcie->dbi + pcie->drvdata->lut_offset;
+	/* Disable LDR zero */
+	iowrite32(0, pcie->lut + PCIE_LUT_LDR(0));
 
 	if (!ls_pcie_is_bridge(pcie))
 		return -ENODEV;
 
+	if (of_device_is_compatible(pdev->dev.of_node, "fsl,ls2085a-pcie") ||
+	of_device_is_compatible(pdev->dev.of_node, "fsl,ls2080a-pcie")) {
+		int len;
+		const u32 *prop;
+		struct device_node *np;
+
+		np = pdev->dev.of_node;
+		prop = (u32 *)of_get_property(np, "available-stream-ids", &len);
+		if (prop) {
+			pcie->avail_streamids = prop;
+			pcie->streamid_index = len/sizeof(u32);
+		} else
+			dev_err(&pdev->dev, "PCIe endpoint partitioning not possible\n");
+	}
+
 	ret = ls_add_pcie_port(&pcie->pp, pdev);
 	if (ret < 0)
 		return ret;
diff --git a/drivers/pci/host/pci-layerscape.h b/drivers/pci/host/pci-layerscape.h
index 8c0642a..e90e114 100644
--- a/drivers/pci/host/pci-layerscape.h
+++ b/drivers/pci/host/pci-layerscape.h
@@ -7,26 +7,7 @@
 #ifndef _PCI_LAYERSCAPE_H
 #define _PCI_LAYERSCAPE_H
 
-/* PEX1/2 Misc Ports Status Register */
-#define SCFG_PEXMSCPORTSR(pex_idx)	(0x94 + (pex_idx) * 4)
-#define LTSSM_STATE_SHIFT	20
-#define LTSSM_STATE_MASK	0x3f
-#define LTSSM_PCIE_L0		0x11 /* L0 state */
-
-/* PEX LUT registers */
-#define PCIE_LUT_BASE		0x80000
-#define PCIE_LUT_PEXLSR		0x020 /* PEX LUT Status Register */
-#define PCIE_LUT_PEXLCR		0x024 /* PEX LUT Control Register */
-#define PCIE_LUT_DBG		0x7FC /* PEX LUT Debug register */
-#define PCIE_LUT_UDR(n)		(0x800 + (n) * 8)
-#define PCIE_LUT_LDR(n)		(0x804 + (n) * 8)
-#define PCIE_LUT_MASK_ALL	0xffff
-#define PCIE_LUT_DR_NUM		32
-#define PCIE_LUT_ENABLE		(1 << 31)
-
 /* function for setting up stream id to device id translation */
 u32 set_pcie_streamid_translation(struct pci_dev *pdev, u32 devid);
 
-#define PCIE_ATU_NUM		6
-
 #endif /* _PCI_LAYERSCAPE_H */
-- 
2.1.0.27.g96db324

